1D CSR Model

xbojanic opened the pull request:
This PR implements a 1D CSR model.  Currently, this is the ultrarelativistic steady-state wakefield model (eg, eq 19 of 
E. L. Saldin et al, NIMA 398, p. 373-394 (1997), [DOI:10.1016/S0168-9002(97)00822-X](https://doi.org/10.1016/S0168-9002(97)00822-X)).

This includes:

- [x] wake function calculation
  - [x] integrate out singularity value
- [x] beam profile convolution
- [x] apply the force "kick" to the particle momentum.
- [x] benchmark example
- [x] GPU support
- [x] MPI support
- [x] user-facing documentation & controls: inputs & Python

Close #452

Commit:
Commit for adding new wake function scripts and charge deposition script

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
CSR Model C++ Draft

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 reviewed (commented):
Congratulations on posting your first ImpactX pull request!  The changes to the examples (your local tests) probably don't need to appear in this request.  We can add a new test that includes CSR, once the wakefield kick is fully implemented.

Commit:
Revert changes to examples.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Delete examples/cfbend/Wakefield_Functions_Bend_copy.py

Not part of this PR.

cemitch99 committed:
Delete examples/chicane/open_PMD_to_ASCII_copy.py

Not part of this PR.

cemitch99 committed:
Delete examples/fodo/openPMD_to_ASCII.py

Not part of this PR.

Commit:
Created wakefields directory and moved all relevant wake function files into it. Created 2 new files for the wake force momentum kick.

Commit:
Merge remote-tracking branch 'xbojanic/Charge_Deposit_Step' into Charge_Deposit_Step

Created the directory wakefields for all new wakefield scripts.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

xbojanic committed:
Delete src/particles/ChargeBinning.H

xbojanic committed:
Delete src/particles/ChargeBinning.cpp

xbojanic committed:
Delete src/particles/WakeConvolution.H

xbojanic committed:
Delete src/particles/WakeConvolution.cpp

Commit:
Added new paths to files to compile

Commit:
Merge remote-tracking branch 'xbojanic/Charge_Deposit_Step' into Charge_Deposit_Step
Deleted copy files and added paths to new wakefields directory for compiling

Commit:
Added necessary paths to new wakefields directory for compilation

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

xbojanic committed:
Update ImpactX.cpp by deleting lines responsible for plotting

xbojanic committed:
Update ImpactX.cpp by adding  plotting script (commented)

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
Removed variable hardcoding by reading in all function parameters

Commit:
Remove variable hardcoding by reading in function parameters

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 commented on src/particles/wakefields/WakePush.cpp:
```diff
@@ -0,0 +1,81 @@
+
+                //Physical constants and reference quantities
+                amrex::ParticleReal const c0_SI = 2.99792458e8;
+                amrex::ParticleReal const mc_SI = pc.GetRefParticle().mass * c0_SI;
```
```suggestion
                amrex::ParticleReal const mc_SI = pc.GetRefParticle().mass * (ablastr::constant::SI::c);
```

cemitch99 commented on src/particles/wakefields/WakePush.cpp:
```diff
@@ -0,0 +1,81 @@
+                const int np = pti.numParticles();
+
+                //Physical constants and reference quantities
+                amrex::ParticleReal const c0_SI = 2.99792458e8;
```
```suggestion
```

This can be removed by using the internally-stored value of c in the ABLASTR library.

cemitch99 commented on src/particles/wakefields/WakePush.cpp:
```diff
@@ -0,0 +1,81 @@
+                amrex::ParticleReal* const AMREX_RESTRICT part_pz = soa_real[RealSoA::pz].dataPtr(); //Note: Currently for a fixed t
+
+                //Obtain constants for force normalization
+                amrex::ParticleReal const push_consts = 1.0 / (c0_SI * pz_ref_SI);
```
```suggestion
                amrex::ParticleReal const push_consts = 1.0 / ((ablastr::constant::SI::c) * pz_ref_SI);
```

cemitch99 commented on src/particles/wakefields/WakeConvolution.H:
```diff
@@ -0,0 +1,33 @@
+using namespace amrex;
+
+//Constants
+constexpr double e = 1.602e-19;  // Elementary charge [C]
```
This can be replaced by using the elementary charge in `ablastr::constant::SI::q_e`

cemitch99 commented on src/particles/wakefields/WakeConvolution.H:
```diff
@@ -0,0 +1,33 @@
+
+//Constants
+constexpr double e = 1.602e-19;  // Elementary charge [C]
+constexpr double c = 3.0e8;  //Speed of light [m/s]
```
This can be replaced by using the speed of light in `ablastr::constant::SI::c`

cemitch99 commented on src/particles/wakefields/WakeConvolution.H:
```diff
@@ -0,0 +1,33 @@
+//Constants
+constexpr double e = 1.602e-19;  // Elementary charge [C]
+constexpr double c = 3.0e8;  //Speed of light [m/s]
+constexpr double m_e = 9.109e-31;  //Electron mass [kg]
```
The same goes for many of these constants:
https://warpx.readthedocs.io/en/23.11/_static/doxyhtml/namespaceablastr_1_1constant_1_1_s_i.html

cemitch99 commented on src/particles/wakefields/ChargeBinning.H:
```diff
@@ -0,0 +1,17 @@
+
+//Constants
+constexpr Real e_charge = 1.602e-19; //Elementary charge in [C]
+
```
Can be replaced by ablastr::constant::SI::q_e

xbojanic committed:
Update src/particles/wakefields/WakePush.cpp

Co-authored-by: Chad Mitchell <46825199+cemitch99@users.noreply.github.com>

xbojanic committed:
Update src/particles/wakefields/WakePush.cpp

Co-authored-by: Chad Mitchell <46825199+cemitch99@users.noreply.github.com>

xbojanic committed:
Update src/particles/wakefields/WakePush.cpp

Co-authored-by: Chad Mitchell <46825199+cemitch99@users.noreply.github.com>

Commit:
Access variables directly using ablastr

Commit:
Aceessing constant variables using ablastr
Merge remote-tracking branch 'xbojanic/Charge_Deposit_Step' into Charge_Deposit_Step

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

xbojanic committed:
Update WakePush.cpp

cemitch99 commented on src/ImpactX.cpp:
```diff
@@ -199,6 +208,100 @@ namespace impactx {
+                        /*
+                        else if constexpr (std::is_same_v<std::decay_t<decltype(element)>, ExactSbend>) //Currently internal calculation for m_rc
+                        {
+                            R = element.m_rc;
```
This case (the `ExactSBend`) will cause a problem because `m_rc` is not a member variable.  Instead, `rc` is computed internally from other parameters.  For now, we could add a comment that this case will be addressed later.  We can make a simple change to ExactSBend to address this.

ax3l replied:
Adding a TODO now in the code.

ax3l replied:
Actually, I can implement it real quick.

Commit:
Added changes to python binding modules to include wakefield python test.

Commit:
Added python bindings for wakefield python test

Commit:
Merging changes for adding python bindings for wakefield python test
Merge remote-tracking branch 'xbojanic/Charge_Deposit_Step' into Charge_Deposit_Step

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l committed:
CMake: `ImpactX_FFT` Option

Modernize FFT build.

ax3l committed:
CMake: Ensure non-FFT Builds Still Work

Guard FFTW logic and throw runtime errors if used but not built.

ax3l committed:
Docs: FFTW Installation

xbojanic committed:
Merge remote-tracking branch 'mainline/development' into Charge_Deposit_Step

xbojanic committed:
Added wake copy test with binning functions

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

xbojanic committed:
Add binning python bindings to all relevant files

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l commented on src/ImpactX.cpp:
```diff
@@ -29,8 +31,14 @@
 #include <memory>
 
+#include <vector>
+#include <fftw3.h>
```
Not needed in this file :)
```suggestion

```

xbojanic committed:
Added working wakefield plotting test script in python

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

xbojanic committed:
Update test_wake_copy.py

xbojanic committed:
Updated wakefield files giving expected emittance growth

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

xbojanic committed:
Added CSR scripts to run and analyze CSR wake effects in chicane

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

xbojanic committed:
Update analysis_chicane_csr.py

xbojanic committed:
Added CSR tests to examples/CMakeLists.txt

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update README.rst

Added CSR example description.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update CMakeLists.txt

Modify to lowercase.

xbojanic committed:
Added CSR inputs

xbojanic committed:
Added CSR inputs

xbojanic committed:
Delete tests/python/test_wake.py

xbojanic committed:
Delete tests/python/test_wake_copy.py

xbojanic committed:
Added wakefield plotting script

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

xbojanic committed:
Removed unused variables from wakefields files

xbojanic committed:
Edited files for checks

xbojanic committed:
Merge branch 'development' into Charge_Deposit_Step

xbojanic committed:
Addressing check errors in wake files

xbojanic committed:
Edited WakePush to allow force indexing

xbojanic committed:
Update WakePush.cpp

xbojanic committed:
Update WakePush.H

xbojanic committed:
Update WakePush.cpp

xbojanic committed:
Ensure test_wake.py only runs for MPI turned off

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

xbojanic committed:
Update test_wake.py

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

xbojanic committed:
Delete tests/python/test_wake.py

xbojanic committed:
Update CMakeLists.txt

xbojanic committed:
Update test_charge_deposition.py

xbojanic committed:
Update test_charge_deposition.py

xbojanic committed:
Update test_charge_deposition.py

xbojanic committed:
Update CMakeLists.txt

cemitch99 commented on src/particles/wakefields/WakeConvolution.cpp:
```diff
@@ -0,0 +1,152 @@
+#include "initialization/InitDistribution.H"
+
+#ifdef ImpactX_USE_FFT
+#include <fftw3.h> //Fastest Fourier Transform in the West
```
This line causes a problem during CI testing.  Now that access to FFTW is enabled elsewhere, it's not clear if this line is needed, or how it should be modified.  Suggest trying without this line.

xbojanic replied:
Removing this line causes compilation errors locally for unknown type names associated with the FFTW variable types.

cemitch99 replied:
Understood.  It looks like this is the main problem left.  Only 3 of the tests are failing, and in 2 cases this is because the fftw header file isn't located during the build.  We may need to consult with Axel on this part.

ax3l replied:
I will help here. On GPU (CUDA/HIP), we will use GPU-accelerated FFTs instead of FFTW. We have already wrappers in ABLASTR for this, which we can use here, too.

ax3l replied:
First updates pushed today.
We will pull in an update to ABLASTR/WarpX here once this PR is merged: https://github.com/ECP-WarpX/WarpX/pull/5112

ax3l committed:
Simplify FFTs: R2C and C2R

ax3l committed:
CMake: Fix ABLASTR FFTW Superbuild

Forgot to forward FFT request.

ax3l committed:
FFTW -> AnyFFT from ABLASTR

This requires a PR to WarpX/ABLASTR that we will post separately.

ax3l committed:
C++ Files: Copyright Headers

ax3l commented on src/particles/wakefields/WakeConvolution.cpp:
```diff
@@ -0,0 +1,167 @@
+*/
+
+//Step Function
+double unit_step(double s)
```
As discussed: replacing here and in the other new files all `double` with `amrex::Real` allows us to compile for single precision (non-default).

ax3l commented on src/particles/wakefields/WakeConvolution.cpp:
```diff
@@ -0,0 +1,167 @@
+
+    //Allocate memory for FFT inputs and outputs
+    using ablastr::math::anyfft::Complex;
+    double *in1 = (double*) malloc(sizeof(double) * n); //Allocate memory for 'n' real numbers for inputs and complex outputs
+    double *in2 = (double*) malloc(sizeof(double) * n);
+    Complex *out1 = (Complex*) malloc(sizeof(Complex) * n);
+    Complex *out2 = (Complex*) malloc(sizeof(Complex) * n);
+    Complex *conv_result = (Complex*) malloc(sizeof(Complex) * n);
+    double *out3 = (double*) malloc(sizeof(double) * n);
```
We will as a last  step replace these allocations with `amrex::Gpu::DeviceVector<Real>` and copy in the data (host-device) from the preparatory calculations.

ax3l commented on src/main.cpp:
```diff
@@ -1,12 +1,13 @@
 /* Copyright 2022-2023 The Regents of the University of California, through Lawrence
- *           Berkeley National Laboratory (subject to receipt of any required
- *           approvals from the U.S. Dept. of Energy). All rights reserved.
+ * Berkeley National Laboratory (subject to receipt of any required
+ * approvals from the U.S. Dept. of Energy). All rights reserved.
  *
  * This file is part of ImpactX.
  *
```
Let us undo all changes in this file.

ax3l commented on src/particles/ImpactXParticleContainer.H:
```diff
@@ -19,6 +19,7 @@
 #include <AMReX_ParIter.H>
 #include <AMReX_Particles.H>
 #include <AMReX_ParticleTile.H>
+#include <AMReX_REAL.H> //added line
```
```suggestion
#include <AMReX_REAL.H>
```

ax3l commented on src/particles/wakefields/ChargeBinning.H:
```diff
@@ -0,0 +1,23 @@
+
+#include "../ImpactXParticleContainer.H"
+
+using namespace amrex;
```
We cannot use namespace in header files, since this would propagate to all files that include the header and then cause collisions.
```suggestion
```

Please prefix AMReX types in this file explicitly with `amrex::`, e.g., `amrex::Real`.

ax3l commented on src/particles/wakefields/ChargeBinning.cpp:
```diff
@@ -0,0 +1,109 @@
+ * License: BSD-3-Clause-LBNL
+ */
+#include "ChargeBinning.H"
+#include "../ImpactXParticleContainer.H" //Includes all necessary AMReX headers
```
Relative to `src/`:
```suggestion
#include "particles/ImpactXParticleContainer.H"
```

ax3l commented on src/particles/wakefields/ChargeBinning.H:
```diff
@@ -0,0 +1,23 @@
+
+using namespace amrex;
+
+//Function to calculate charge density profile
```
we need to add `namespace impactx::particles::wakefields { ... }` to this file.

ax3l commented on src/particles/wakefields/WakeConvolution.H:
```diff
@@ -0,0 +1,37 @@
+#include "particles/ImpactXParticleContainer.H"
+#include "initialization/InitDistribution.H"
+
+
```
we need to add `namespace impactx::particles::wakefields { ... }` to this file.

ax3l commented on src/particles/wakefields/WakePush.H:
```diff
@@ -0,0 +1,23 @@
+#include <vector>
+#include "../ImpactXParticleContainer.H"
+
+namespace impactx::wakepush
```
mirror the names of the directories this file is in:
```suggestion
namespace impactx::particles::wakefields
```

ax3l commented on src/particles/wakefields/WakePush.cpp:
```diff
@@ -0,0 +1,88 @@
+#include <AMReX_REAL.H>
+#include <AMReX_SPACE.H>
+
+namespace impactx::wakepush
```
```suggestion
namespace impactx::particles::wakefields
```

ax3l commented on src/python/WakeConvolution.cpp:
```diff
@@ -0,0 +1,50 @@
+#include "pyImpactX.H"
```
Please add the copyright header comment block (see, e.g., `ChargeBinning.H`

xbojanic committed:
Updated variable double to Real type

ax3l commented on src/ImpactX.cpp:
```diff
@@ -17,6 +18,7 @@
 #include "particles/spacecharge/GatherAndPush.H"
 #include "particles/spacecharge/PoissonSolve.H"
 #include "particles/transformation/CoordinateTransformation.H"
+#include "particles/elements/Sbend.H"
```
This should not be explicitly used here.
```suggestion
```

ax3l commented on src/ImpactX.cpp:
```diff
@@ -199,6 +215,95 @@ namespace impactx {
 
+                    //CSR Wakefield Response
+
+                    bool element_has_csr = false; // Updates to true for example with bend element
```
We need to generalize this, e.g., as an attribute to the elements that have CSR or as a small map/list of elements that have CSR, stored external to them.

I assume we will have multiple wakefield models for various elements later on?

ax3l commented on src/ImpactX.cpp:
```diff
@@ -199,6 +215,95 @@ namespace impactx {
+                    //CSR Wakefield Response
+
+                    bool element_has_csr = false; // Updates to true for example with bend element
+                    amrex::Real R = 0.0; // Updates for bend element rc
+
+                    //Define lambda function inside of std::visit
+                    std::visit([&R, &element_has_csr](auto &&element)
+                    {
+                        if constexpr (std::is_same_v<std::decay_t<decltype(element)>, Sbend>)
+                        {
+                            R = element.m_rc;
+                            R = std::abs(R);
+                            element_has_csr = true;
+                        }
+                        else if constexpr (std::is_same_v<std::decay_t<decltype(element)>, CFbend>)
+                        {
+                            R = element.m_rc;
+                            R = std::abs(R);
+                            element_has_csr = true;
+                        }
+
```
All these lines I would put inside a function call, so that `ImpactX.cpp` stays focused on overall control logic, and this then is called in a CSR-effect function that checks if CSR shall be calculated for an element and how.

xbojanic committed:
Update src/particles/ImpactXParticleContainer.H

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

xbojanic committed:
Update main.cpp

xbojanic committed:
Update src/particles/wakefields/ChargeBinning.cpp

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

xbojanic committed:
Update src/particles/wakefields/WakePush.H

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

xbojanic committed:
Update src/particles/wakefields/WakePush.cpp

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

xbojanic committed:
Update WakePush.cpp

xbojanic committed:
Update src/ImpactX.cpp

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

xbojanic committed:
Minor edits to import statements and declarations in wakefield files

xbojanic committed:
Encapsulate all wakefield header files in wakefields namespace

xbojanic committed:
Added namespace changes to .cpp files and ImpactX.cpp

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l committed:
Add Docstrings, Formatting

Co-authored-by: Alex Bojanich <abojanich@lbl.gov>

ax3l committed:
Merge remote-tracking branch 'mainline/development' into Charge_Deposit_Step

ax3l committed:
Fix Signed/Unsigned Warning

xbojanic committed:
Added template functions for handling CSR bend elements

ax3l commented on src/initialization/InitDistribution.H:
```diff
@@ -17,6 +17,8 @@
 
 #include <utility>  // for std::move
 
+// Declare the external variable
+extern amrex::ParticleReal bunch_charge;
```
We don't use global variables - let me know if you still need this and we'll pass it around or get it from ParmParse or the particle container :)
```suggestion
```

ax3l commented on src/initialization/InitDistribution.cpp:
```diff
@@ -28,6 +28,8 @@
 #include <type_traits>
 #include <variant>
 
+//Global declaration
+amrex::ParticleReal bunch_charge = 0.0;
```
Please see above.
```suggestion
```

ax3l commented on src/initialization/InitDistribution.cpp:
```diff
@@ -219,7 +221,7 @@ namespace impactx
         pp_dist.get("kin_energy", kin_energy);
 
-        amrex::ParticleReal bunch_charge = 0.0;  // Bunch charge (C)
+        //amrex::ParticleReal bunch_charge = 0.0;  // Bunch charge (C) - Define externally to make global variable
```
Please see above
```suggestion
        amrex::ParticleReal bunch_charge = 0.0;  // Bunch charge (C)
```

xbojanic committed:
Remove global variable declaration for charge

xbojanic committed:
Added last function for calculating mean transverse position. Created singular function call for wakefield effects. Updated function that accesses bend elements.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l committed:
Update Doxygen Argument Name

ax3l committed:
Cleaning: Leftover Changes

Remove leftover changes that change files we do not touch.
Minimize the diff of this PR.

ax3l commented on src/particles/wakefields/ExecuteWakefield.H:
```diff
@@ -0,0 +1,122 @@
+        if (csr && element_has_csr)
+        {
+            // Measure beam size, extract the min, max of particle positions
+            auto const [x_min, y_min, t_min, x_max, y_max, t_max] =
```
resolved.

ax3l commented on src/particles/wakefields/ExecuteWakefield.H:
```diff
@@ -0,0 +1,122 @@
+        if (csr && element_has_csr)
+        {
+            // Measure beam size, extract the min, max of particle positions
+            auto const [x_min, y_min, t_min, x_max, y_max, t_max] =
```
resolved.

ax3l commented on src/particles/wakefields/ExecuteWakefield.H:
```diff
@@ -0,0 +1,122 @@
+        if (csr && element_has_csr)
+        {
+            // Measure beam size, extract the min, max of particle positions
+            auto const [x_min, y_min, t_min, x_max, y_max, t_max] =
```
resolved.

ax3l commented on src/particles/wakefields/ExecuteWakefield.H:
```diff
@@ -0,0 +1,122 @@
+        if (csr && element_has_csr)
+        {
+            // Measure beam size, extract the min, max of particle positions
+            auto const [x_min, y_min, t_min, x_max, y_max, t_max] =
```
resolved.

ax3l commented on src/particles/wakefields/ExecuteWakefield.H:
```diff
@@ -0,0 +1,122 @@
+
+        // Call the CSR bend function
+        impactx::particles::wakefields::CSRBendElement(R, element_has_csr, element_variant);
+        //impactx::particles::wakefields::HandleElementVariant(R, element_has_csr, element_variant);
```
resolved.

ax3l commented on src/particles/wakefields/WakePush.cpp:
```diff
@@ -0,0 +1,97 @@
+                {
+                    // Access SoA Real data
+                    // amrex::ParticleReal & AMREX_RESTRICT x = part_x[i];
+                    // amrex::ParticleReal & AMREX_RESTRICT y = part_y[i];
```
resolved.

ax3l commented on src/particles/wakefields/WakePush.cpp:
```diff
@@ -0,0 +1,97 @@
+                    amrex::ParticleReal & AMREX_RESTRICT z = part_z[i];
+
+                    // amrex::ParticleReal & AMREX_RESTRICT px = part_px[i];
+                    // amrex::ParticleReal & AMREX_RESTRICT py = part_py[i];
```
resolved.

ax3l committed:
More Cleaning

ax3l committed:
Merge remote-tracking branch 'mainline/development' into Charge_Deposit_Step

ax3l committed:
GPU Support

ax3l commented on src/particles/wakefields/WakePush.cpp:
```diff
@@ -0,0 +1,84 @@
+                // Access data from StructOfArrays (soa)
+                auto& soa_real = pti.GetStructOfArrays().GetRealData();
+
+                amrex::ParticleReal* const AMREX_RESTRICT part_z = soa_real[RealSoA::z].dataPtr(); // Note: Currently for a fixed t
+                amrex::ParticleReal* const AMREX_RESTRICT part_pz = soa_real[RealSoA::pz].dataPtr(); // Note: Currently for a fixed t
```
Is that comment correct or copied? I think we operate on fixed `s` here because we did no s-to-t transform.

cemitch99 replied:
Chiming in here:  Yes, there should be no s-to-t transform.  It looks like this was copied.

ax3l replied:
Let's then rename to:
```suggestion
                amrex::ParticleReal* const AMREX_RESTRICT part_t = soa_real[RealSoA::t].dataPtr();
                amrex::ParticleReal* const AMREX_RESTRICT part_pt = soa_real[RealSoA::pt].dataPtr();
```

ax3l committed:
CMake: Comment in CSR Test

ax3l commented on examples/CMakeLists.txt:
```diff
@@ -890,3 +890,23 @@ add_impactx_test(spectrometer.py
+add_impactx_test(chicane_csr
+    examples/chicane/input_chicane_csr.in
+      OFF  # ImpactX MPI-parallel
+      OFF  # ImpactX Python interface
```
```suggestion
```

ax3l commented on examples/CMakeLists.txt:
```diff
@@ -890,3 +890,23 @@ add_impactx_test(spectrometer.py
+add_impactx_test(chicane_CSR.py
+    examples/chicane/run_chicane_csr.py
+      OFF  # ImpactX MPI-parallel
+      ON   # ImpactX Python interface
```
```suggestion
```

ax3l committed:
add_impactx_test: new simpler interface

ax3l commented on examples/chicane/analysis_chicane_csr.py:
```diff
@@ -0,0 +1,97 @@
+    epstrms = beam.cov(ddof=0)
+    emittance_x = (sigx**2 * sigpx**2 - epstrms["position_x"]["momentum_x"] ** 2) ** 0.5
+    emittance_y = (sigy**2 * sigpy**2 - epstrms["position_y"]["momentum_y"] ** 2) ** 0.5
+    emittance_t = (sigt**2 * sigpt**2 - epstrms["position_t"]["momentum_t"] ** 2) ** 0.5
```
Runtime  warning:
```
/Users/runner/work/impactx/impactx/examples/chicane/analysis_chicane_csr.py:31: RuntimeWarning: invalid value encountered in scalar subtract
  emittance_t = (sigt**2 * sigpt**2 - epstrms["position_t"]["momentum_t"] ** 2) ** 0.5
```
We should check if `position_t` and `momentum_t` all have valid values.

I think I saw some potential issues in `WakePush` yesterday where `pt` got invalid - we might need some extra guarding for unexpected/out-of-range values there.

ax3l commented on examples/chicane/run_chicane_csr.py:
```diff
@@ -0,0 +1,83 @@
+
+#   particle bunch
+distr = distribution.Gaussian(
+    lambdaX=2.2951017632e-5,
+    lambdaY=1.3084093142e-5,
+    lambdaT=5.5555553e-8,
+    lambdaPx=1.598353425e-6,
+    lambdaPy=2.803697378e-6,
+    lambdaPt=2.000000000e-6,
+    muxpx=0.933345606203060,
+    muypy=0.933345606203060,
+    mutpt=0.999999961419755,
```
A lot of the distribution parameters and the accelerator lattice params `rc`, `psi`, `lb` and `ds` are "magic constants".

Since we are in Python here, can we calculate them to make them self-describing?

cemitch99 replied:
These parameters are set by the Berlin-Zeuthen chicane design.  For a general 4-bend chicane, they are related to the initial beam parameters and to the required final compression factor, and the relations are rather involved.  For testing with CSR, there are some approximate formulas for the predicted emittance growth, but this adds additional complications.  I suggest that this can be revisited later, with low priority.

ax3l replied:
Agreed, thanks for clarifying.

ax3l committed:
Merge remote-tracking branch 'mainline/development' into Charge_Deposit_Step

ax3l committed:
MPI Support

ax3l commented on src/particles/wakefields/ExecuteWakefield.H:
```diff
@@ -0,0 +1,121 @@
+            amrex::Real* dptr_convoluted_wakefield = convoluted_wakefield.data();
+            impactx::particles::wakefields::convolve_fft(slopes, wake_function, bin_size, convoluted_wakefield, padding_factor);
+
+            // FIXME: MPI BCast convoluted_wakefield to every MPI rank
```
resolved

ax3l commented on src/particles/wakefields/ExecuteWakefield.H:
```diff
@@ -0,0 +1,121 @@
+            // Call charge deposition function
+            impactx::particles::wakefields::DepositCharge1D(particle_container, charge_distribution, bin_min, bin_size, is_unity_particle_weight);
+
+            // FIXME: MPI reduce to rank zero, then the rest on IO processor, then BCast result
```
resolved

xbojanic committed:
Add .rst CSR effects documentation. Ensure momentum update values are valid.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l commented on docs/source/usage/parameters.rst:
```diff
@@ -590,6 +590,19 @@ Multigrid-specific numerical options:
+CSR effects are included in the simulation for bend lattice elements such as Sbend and CFbend. These effects are critical in accurately modeling the wakefields generated due to the interaction of particles with the radiated field during bending in synchrotron radiation.
+
+* ``algo.csr`` (boolean, optional, default: false)
+Whether to calculate CSR effects.
+CSR calculations involve several steps, including charge deposition, mean transverse position calculation, wakefield generation, and convolution, all of which are handled within the CSR bending process.
+
+* ``algo.csr_bins`` (integer, optional, default: 150)
+The number of bins used for the CSR calculations along the longitudinal direction. Increasing the number of bins can lead to more accurate wakefield resolution at the cost of higher computational expense.
+
+.. note::
+
+CSR effects are only calculated for lattice elements that include bending, such as Sbend and CFbend.
```
These spaces are needed so Sphinx/rst formats this correctly (bullets & blocks):
```suggestion
    Whether to calculate CSR effects.
    CSR calculations involve several steps, including charge deposition, mean transverse position calculation, wakefield generation, and convolution, all of which are handled within the CSR bending process.

* ``algo.csr_bins`` (integer, optional, default: 150)
    The number of bins used for the CSR calculations along the longitudinal direction. Increasing the number of bins can lead to more accurate wakefield resolution at the cost of higher computational expense.

.. note::

   CSR effects are only calculated for lattice elements that include bending, such as Sbend and CFbend.
```

ax3l committed:
Sphinx/rst Formatting

ax3l commented on docs/source/usage/parameters.rst:
```diff
@@ -590,6 +590,19 @@ Multigrid-specific numerical options:
 
+CSR effects are included in the simulation for bend lattice elements such as Sbend and CFbend. These effects are critical in accurately modeling the wakefields generated due to the interaction of particles with the radiated field during bending in synchrotron radiation.
+
+* ``algo.csr`` (boolean, optional, default: false)
+    Whether to calculate CSR effects.
+    CSR calculations involve several steps, including charge deposition, mean transverse position calculation, wakefield generation, and convolution, all of which are handled within the CSR bending process.
+
+* ``algo.csr_bins`` (integer, optional, default: 150)
```
Verbatim formatting as in other parameters:
```suggestion
* ``algo.csr`` (``boolean``, optional, default: ``false``)
    Whether to calculate CSR effects.
    CSR calculations involve several steps, including charge deposition, mean transverse position calculation, wakefield generation, and convolution, all of which are handled within the CSR bending process.

* ``algo.csr_bins`` (``integer`, optional, default: ``150``)
```

ax3l committed:
Verbatim Options

ax3l commented on src/particles/wakefields/WakePush.cpp:
```diff
@@ -0,0 +1,95 @@
+                {
+                    // Access SoA Real data
+                    amrex::ParticleReal const & AMREX_RESTRICT z = part_z[i];
+                    amrex::ParticleReal & AMREX_RESTRICT pz = part_pz[i];
```
```suggestion
                    amrex::ParticleReal const & AMREX_RESTRICT t = part_t[i];
                    amrex::ParticleReal & AMREX_RESTRICT pt = part_pt[i];
```

ax3l commented on src/particles/wakefields/WakePush.cpp:
```diff
@@ -0,0 +1,95 @@
+
+                    // Update longitudinal momentum with the convoluted wakefield force
+                    amrex::Real const lower_bound = padding_factor * 2 * bin_min;
+                    int const idx = static_cast<int>((z - lower_bound) / bin_size); // Find index position along z
```
```suggestion
                    int const idx = static_cast<int>((t - lower_bound) / bin_size); // Find index position along t
```

ax3l commented on src/particles/wakefields/WakePush.cpp:
```diff
@@ -0,0 +1,95 @@
+                    if (std::isfinite(F_L))
+                    {
+                        // Update longitudinal momentum
+                        pz -= push_consts * slice_ds * F_L;
```
```suggestion
                        pt -= push_consts * slice_ds * F_L;
```

ax3l commented on src/particles/wakefields/WakePush.H:
```diff
@@ -0,0 +1,42 @@
+{
+    /** Function to update particle momentum using wake force kick
+     *
+     * @param[inout] pc
+     * @param[in] convoluted_wakefield
+     * @param[in] slice_ds
+     * @param[in] bin_size
+     * @param[in] bin_min
+     * @param[in] padding_factor
```
@xbojanic please complete the documentation of these arguments (description & unit for each)

ax3l replied:
added :)

cemitch99 commented on docs/source/usage/parameters.rst:
```diff
@@ -590,6 +590,19 @@ Multigrid-specific numerical options:
 Coherent Synchrotron Radiation (CSR)
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
+CSR effects are included in the simulation for bend lattice elements such as Sbend and CFbend. These effects are critical in accurately modeling the wakefields generated due to the interaction of particles with the radiated field during bending in synchrotron radiation.
```
```suggestion
CSR effects are included in the simulation for bend lattice elements such as Sbend and CFbend. These effects are critical in accurately modeling the wakefields generated due to the interaction of particles with the synchrotron radiation field generated by the beam during bending.  Currently, this is the 1D ultrarelativistic steady-state wakefield model (eq 19 of
E. L. Saldin et al, NIMA 398, p. 373-394 (1997), [DOI:10.1016/S0168-9002(97)00822-X]).
```

cemitch99 commented on docs/source/usage/parameters.rst:
```diff
@@ -590,6 +590,19 @@ Multigrid-specific numerical options:
+
+* ``algo.csr`` (``boolean``, optional, default: ``false``)
+    Whether to calculate CSR effects.
+    CSR calculations involve several steps, including charge deposition, mean transverse position calculation, wakefield generation, and convolution, all of which are handled within the CSR bending process.
```
```suggestion
    CSR calculations involve several steps, including charge deposition, wakefield generation, and convolution, all of which are handled within the CSR bending process.
```

cemitch99 replied:
The mean transverse position calculation is not used for CSR calculations.

cemitch99 commented on examples/chicane/README.rst:
```diff
@@ -72,3 +72,56 @@ You can run the following script to visualize the beam evolution over time:
+A 5 GeV electron bunch with normalized transverse rms emittance of 1 um undergoes longitudinal compression by a factor of 10 in a standard 4-bend chicane.
+The rms pulse length should decrease by the compression factor (10).
+
+An ultrarelativistic model of steady-state CSR in the bending dipoles is included, resulting in a horizontal emittance growth of 19%.
```
```suggestion
An ultrarelativistic model of steady-state CSR in the bending dipoles is included, resulting in a horizontal emittance growth of 19%.  Note that this value is smaller than the horizontal emittance growth of 57% that is obtained when transient (bend-entry and -exit) effects are included.
```

xbojanic committed:
Update src/particles/wakefields/WakePush.cpp

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

xbojanic committed:
Update src/particles/wakefields/WakePush.cpp

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

xbojanic committed:
Update src/particles/wakefields/WakePush.cpp

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

xbojanic committed:
Update src/particles/wakefields/WakePush.cpp

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

xbojanic committed:
Update docs/source/usage/parameters.rst

Co-authored-by: Chad Mitchell <46825199+cemitch99@users.noreply.github.com>

xbojanic committed:
Update docs/source/usage/parameters.rst

Co-authored-by: Chad Mitchell <46825199+cemitch99@users.noreply.github.com>

xbojanic committed:
Update examples/chicane/README.rst

Co-authored-by: Chad Mitchell <46825199+cemitch99@users.noreply.github.com>

xbojanic committed:
Add variable descriptions to WakePush

ax3l commented on examples/CMakeLists.txt:
```diff
@@ -891,7 +891,22 @@ add_impactx_test(spectrometer.py
+    examples/chicane/analysis_chicane_csr.py
+    OFF  # no plot script yet
+)
+add_impactx_test(chicane_CSR.py
```
Just for consistency in naming:
```suggestion
add_impactx_test(chicane_csr.py
```

cemitch99 replied:
It looks like something broke due to the recent GPU+MPI support changes.  8/16 of the tests now fail, and the output diagnostics grow out of control.  The last verified working version was at the time of **Aug. 14 commit**.  The problem does not appear to be due to recent changes to WakePush.cpp.

ax3l replied:
On CI, only
```
The following tests FAILED:
	190 - chicane_csr.analysis (Failed)
	192 - chicane_CSR.py.analysis (Failed)
```
seem to fail. But rolling back to August 14th had these tests commented out, I'll check if they passed back then.

ax3l replied:
I rolled back to August 14 locally and `chicane_csr.analysis` passes.

ax3l replied:
Ok found it, I broke it in the `GPU support` commit
https://github.com/ECP-WarpX/impactx/pull/627/commits/610689ae206173bc9bd0a72711cae1f475497998

ax3l replied:
I found a few smaller issues and reverted locally a few key parts in that commit, but could not yet nail down the root cause of the issue. we could printf tomorrow the array values at each step of the wakefield calculation to see which one goes off.

ax3l commented on src/particles/wakefields/ChargeBinning.cpp:
```diff
@@ -0,0 +1,191 @@
+        int const nlevs = myspc.finestLevel();
+        for (int lev = 0; lev <= nlevs; ++lev)
+        {
+            #ifdef AMREX_USE_OMP
+            #pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+            #endif
```
On CPU, this should be a race condition because `sum_*` arrays are not defined thread local.
```suggestion
```

ax3l commented on src/particles/wakefields/WakeConvolution.cpp:
```diff
@@ -0,0 +1,170 @@
+
+        amrex::Real const s0 = (0.169_rt * std::pow(a, 1.79_rt) * std::pow(g, 0.38_rt)) / std::pow(L, 1.17_rt);
+        amrex::Real const term = std::sqrt(std::abs(s) / s0) * std::exp(-std::sqrt(std::abs(s) / s0));
+        return (4 * impactx::particles::wakefields::Z0 * ablastr::constant::SI::c * s0 * unit_step(s)) / (amrex::Real(M_PI) * std::pow(a, 4)) * term;
```
Avoid implicit type cast from int to float/double:
```suggestion
        return (4_rt * impactx::particles::wakefields::Z0 * ablastr::constant::SI::c * s0 * unit_step(s)) / (amrex::Real(M_PI) * std::pow(a, 4)) * term;
```

ax3l commented on src/particles/wakefields/WakePush.cpp:
```diff
@@ -0,0 +1,95 @@
+                    else
+                    {
+                        // Handle unexpected values: log warning and skip momentum update
+                        std::cerr << "Warning: Invalid or out-of-range values detected." << std::endl;
```
In a final cleanup, this should be guarded for debug-mode only. It is generally costly and should be a printf on GPU.

ax3l commented on src/particles/wakefields/WakeConvolution.cpp:
```diff
@@ -0,0 +1,170 @@
+        int const wake_func_size = wake_func.size();
+
+        // Length of convolution result
+        int const original_n = beam_profile_slope_size + wake_func_size - 1;  // Output size is n = 2N - 1, where N = size of signals 1,2
```
```suggestion
        int const original_n = beam_profile_slope_size + wake_func_size;  // Output size is n = 2*num_bins - 1, where N = size of signals 1,2 and slope has one element less
```

ax3l replied:
No, needs to be complex size = N/2 + 1 now :
https://www.fftw.org/fftw3_doc/One_002dDimensional-DFTs-of-Real-Data.html

ax3l replied:
Fixed.

ax3l committed:
Bugfix: Vectors Default-Init Zero

For some reason, AMReX device vector does not auto-allocate to
default value type (as `std::vector` does). This fixes isfinite
issues in slope.

ax3l committed:
FFT: Fix FFT Array Sizes

Leftover from C2C rewrite to R2C & C2R FFTs: The array sizes
vary in that case for input and output.
https://www.fftw.org/fftw3_doc/One_002dDimensional-DFTs-of-Real-Data.html

ax3l committed:
Fix: `bin_size`

Upper bound had an off-by-one error for binning.

ax3l committed:
Fix Wake Function Eval and Simplify Padding

ax3l commented on src/ImpactX.cpp:
```diff
@@ -172,6 +174,14 @@ namespace impactx {
         }
 
+        bool csr = false;
+        int csr_bins = 100;
```
Not using the documented default from `parameters.rst`, will fix in Python-binding push:
```suggestion
        int csr_bins = 150;
```

ax3l committed:
Python: User-Facing Bindings & Docs

xbojanic committed:
Updated w_l_csr function to deal with s=0 case

ax3l committed:
Cleaning: Debug and non-GPU Print

- does not work on GPU
- super costly in the hot loop, debug only

cemitch99 commented:
For Windows / MSVC w/o MPI - the chicane_csr test fails because the simulation appears to run without CSR.  No emittance growth is visible.

For Tooling / Clang, HIP / HIP, and macOS / AppleClang - the error is due to an unused variable "cw_size"

cemitch99 committed:
Update WakePush.cpp

Comment unused variable.

cemitch99 committed:
Update WakePush.cpp

Restore cw_size - it is used after all.

cemitch99 committed:
Update WakePush.cpp

Try a conditional declaration of cw_size.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update WakeConvolution.H

Add bin_size to documentation in w_l_csr.

ax3l committed:
FFT & CSR: Earlier User-Facing Abort

and some cleaning

ax3l committed:
Test: Only with FFT

ax3l committed:
Docs: Minor Updates

ax3l committed:
CSR: Support `ExactSbend`

ax3l committed:
CI: FFTW on Windows

ax3l committed:
Include Cleanup

ax3l committed:
Doc: Split Example Table of Content

ax3l commented on src/particles/wakefields/ExecuteWakefield.H:
```diff
@@ -0,0 +1,150 @@
+                impactx::particles::wakefields::DerivativeCharge1D(charge_distribution, slopes, bin_size,
+                                                                   GetNumberDensity); // Use number derivatives for convolution with CSR
+
+                // Construct CSR wake function on 2N support
+                amrex::Gpu::DeviceVector<amrex::Real> wake_function(num_bins*2, 0.0);
+                amrex::Real *const dptr_wake_function = wake_function.data();
+                auto const dR = R;  // for NVCC capture
+                amrex::ParallelFor(num_bins*2, [=] AMREX_GPU_DEVICE(int i) {
+                    if (i < num_bins) {
+                        amrex::Real const s = static_cast<amrex::Real>(i) * bin_size;
+                        dptr_wake_function[i] = impactx::particles::wakefields::w_l_csr(s, dR, bin_size);
+                    }
+                    else if (i > num_bins) {
+                        amrex::Real const s = static_cast<amrex::Real>(i - 2*num_bins) * bin_size;
+                        dptr_wake_function[i] = impactx::particles::wakefields::w_l_csr(s, dR, bin_size);
+                    }
+                });
```
@xbojanic @cemitch99 I cannot yet spot the update for our integration constant at `i=0`. Is that missing?

ax3l replied:
Oh, you treat that directly inside w_l_csr
https://github.com/ECP-WarpX/impactx/pull/627/commits/01422aa1fce947d35073608967577ad4d7640104

ax3l commented on src/particles/wakefields/ExecuteWakefield.H:
```diff
@@ -0,0 +1,150 @@
+/* Copyright 2022-2023 The Regents of the University of California, through Lawrence
```
As a last step of cleaning, I would rename `ExecuteWakefield.H` to `HandleWakefield.H` to match the function inside it.

Let's do this last before merging, so we have all code changes we want already included.

ax3l replied:
Done.

ax3l committed:
Inline Comments: OpenMP

For follow-up PRs on how to add OpenMP support.

ax3l committed:
Rename: `ExecuteWakefield.H` -> `HandleWakefield.H`

To match the function and macros within.

ax3l reviewed (approved):
Thank you for the great work, @xbojanic! Ready to merge from my end :rocket: :sparkles:

ax3l committed:
ChargeBinning: OpenMP

Works, but will not be as fast as it could be.

cemitch99 commented on src/particles/wakefields/HandleWakefield.H:
```diff
@@ -0,0 +1,150 @@
+        impactx::ImpactXParticleContainer& particle_container,
+        T_Element const& element_variant,
+        amrex::Real slice_ds,
+        bool print_wakefield = true
```
```suggestion
        bool print_wakefield = false
```
Let's set this to 'false' by default.

ax3l replied:
Phew, good catch!

cemitch99 commented on src/python/ImpactX.cpp:
```diff
@@ -203,6 +203,26 @@ void init_ImpactX (py::module& m)
+                amrex::ParmParse pp_algo("algo");
+                pp_algo.add("csr_bins", csr_bins);
+            },
+            "Enable or disable Coherent Synchrotron Radiation (CSR) calculations (default: 150)."
```
```suggestion
            "Number of longitudinal bins used for CSR calculations (default: 150)."
```

ax3l committed:
Review: Fix Default & Docstring

Co-authored-by: Chad Mitchell <46825199+cemitch99@users.noreply.github.com>

cemitch99 commented on examples/chicane/analysis_chicane_csr.py:
```diff
@@ -0,0 +1,106 @@
+)
+
+atol = 0.0  # ignored
+rtol = 2.5e1 * num_particles**-0.5  # from random sampling of a smooth distribution
```
```suggestion
rtol = 2.5 * num_particles**-0.5  # from random sampling of a smooth distribution
```
The tolerance is unnecessarily large--tightening this now to settle on a reasonable value.

cemitch99 commented on examples/chicane/analysis_chicane_csr.py:
```diff
@@ -0,0 +1,106 @@
+)
+
+atol = 0.0  # ignored
+rtol = 2.5e1 * num_particles**-0.5  # from random sampling of a smooth distribution
```
```suggestion
rtol = 2.5 * num_particles**-0.5  # from random sampling of a smooth distribution
```
The tolerance is unnecessarily large--tightening this now to settle on a reasonable value.

ax3l committed:
Tighter Tolerances

Co-authored-by: Chad Mitchell <46825199+cemitch99@users.noreply.github.com>

ax3l commented on examples/chicane/analysis_chicane_csr.py:
```diff
@@ -0,0 +1,106 @@
+final = series.iterations[last_step].particles["beam"].to_df()
+
+# compare number of particles
+num_particles = 10000
```
I don't think we need to hard-code the number of particles here.
We can more generically write:
```suggestion
num_particles = len(initial)
```

That way, the input set decides and that is all we need to remember.

ax3l committed:
Analysis: Remove hard-coded number of particles

cemitch99 commented on src/particles/wakefields/HandleWakefield.H:
```diff
@@ -0,0 +1,150 @@
+            // Check convolution output
+            if (print_wakefield && amrex::ParallelDescriptor::IOProcessor())
+            {
+                std::cout << "Convoluted wakefield: ";
```
```suggestion
                std::cout << "Convolved wakefield: ";
```

cemitch99 commented on src/particles/wakefields/HandleWakefield.H:
```diff
@@ -0,0 +1,150 @@
+            if (print_wakefield && amrex::ParallelDescriptor::IOProcessor())
+            {
+                std::cout << "Convoluted wakefield: ";
+                std::ofstream outfile("convoluted_wakefield.txt");
```
```suggestion
                std::ofstream outfile("convolved_wakefield.txt");
```

cemitch99 replied:
```suggestion
                std::ofstream outfile("convolved_wakefield.txt");
```

cemitch99 commented on src/particles/wakefields/HandleWakefield.H:
```diff
@@ -0,0 +1,150 @@
+            {
+                std::cout << "Convoluted wakefield: ";
+                std::ofstream outfile("convoluted_wakefield.txt");
+                for (double const i : convoluted_wakefield) {
```
```suggestion
                for (double const i : convolved_wakefield) {
```

cemitch99 commented on src/particles/wakefields/HandleWakefield.H:
```diff
@@ -0,0 +1,150 @@
+                });
+
+                // Call convolution function
+                convoluted_wakefield = impactx::particles::wakefields::convolve_fft(slopes, wake_function, bin_size);
```
```suggestion
                convolved_wakefield = impactx::particles::wakefields::convolve_fft(slopes, wake_function, bin_size);
```

cemitch99 replied:
```suggestion
                convolved_wakefield = impactx::particles::wakefields::convolve_fft(slopes, wake_function, bin_size);
```

cemitch99 commented on src/particles/wakefields/HandleWakefield.H:
```diff
@@ -0,0 +1,150 @@
+
+            // Broadcast the global wakefield to every MPI rank
+            amrex::ParallelDescriptor::Bcast(
+                convoluted_wakefield.data(),
```
```suggestion
                convolved_wakefield.data(),
```

cemitch99 commented on src/particles/wakefields/HandleWakefield.H:
```diff
@@ -0,0 +1,150 @@
+            // Broadcast the global wakefield to every MPI rank
+            amrex::ParallelDescriptor::Bcast(
+                convoluted_wakefield.data(),
+                convoluted_wakefield.size(),
```
```suggestion
                convolved_wakefield.size(),
```

cemitch99 replied:
I originally thought we should change "convoluted" -> "convolved" for grammatical reasons, but it is used in many places.  I think it's ok to keep as-is.

cemitch99 committed:
Update analysis_chicane_csr.py

Relax tolerance 2.5 -> 3.5.

ax3l commented on src/particles/wakefields/HandleWakefield.H:
```diff
@@ -0,0 +1,150 @@
+     * @param[in] particle_container the particle species container
+     * @param[in] element_variant variant type of the lattice element
+     * @param[in] slice_ds slice spacing along s
+     * @param[in] print_wakefield for debugging: print the wakefield to convoluted_wakefield.txt
```
```suggestion
     * @param[in] print_wakefield for debugging: print the wakefield to convolved_wakefield.txt
```

ax3l commented on src/particles/wakefields/HandleWakefield.H:
```diff
@@ -0,0 +1,150 @@
+            impactx::particles::wakefields::DepositCharge1D(particle_container, charge_distribution, bin_min, bin_size, is_unity_particle_weight);
+
+            // Sum up all partial charge histograms to one MPI process to calculate the global wakefield.
+            // Once calculated, we will distribute convoluted_wakefield back to every MPI process.
```
```suggestion
            // Once calculated, we will distribute convolved_wakefield back to every MPI process.
```

ax3l commented on src/particles/wakefields/HandleWakefield.H:
```diff
@@ -0,0 +1,150 @@
+                amrex::ParallelDescriptor::Communicator()
+            );
+
+            amrex::Gpu::DeviceVector<amrex::Real> convoluted_wakefield;
```
```suggestion
            amrex::Gpu::DeviceVector<amrex::Real> convolved_wakefield;
```

ax3l commented on src/particles/wakefields/HandleWakefield.H:
```diff
@@ -0,0 +1,150 @@
+            }
+
+            // Call function to kick particles with wake
+            impactx::particles::wakefields::WakePush(particle_container, convoluted_wakefield, slice_ds, bin_size, t_min);
```
```suggestion
            impactx::particles::wakefields::WakePush(particle_container, convolved_wakefield, slice_ds, bin_size, t_min);
```

ax3l commented on src/particles/wakefields/WakePush.H:
```diff
@@ -0,0 +1,39 @@
+    /** Function to update particle momentum using wake force kick
+     *
+     * @param[inout] pc the particle species container
+     * @param[in] convoluted_wakefield wake functions in [V*pc/mm]
```
```suggestion
     * @param[in] convolved_wakefield wake functions in [V*pc/mm]
```

ax3l commented on src/particles/wakefields/WakePush.H:
```diff
@@ -0,0 +1,39 @@
+     */
+    void WakePush (
+        ImpactXParticleContainer & pc,
+        amrex::Gpu::DeviceVector<amrex::Real> const & convoluted_wakefield,
```
```suggestion
        amrex::Gpu::DeviceVector<amrex::Real> const & convolved_wakefield,
```

ax3l commented on src/particles/wakefields/WakePush.cpp:
```diff
@@ -0,0 +1,100 @@
+{
+    void WakePush (
+        ImpactXParticleContainer & pc,
+        amrex::Gpu::DeviceVector<amrex::Real> const & convoluted_wakefield,
```
```suggestion
        amrex::Gpu::DeviceVector<amrex::Real> const & convolved_wakefield,
```

ax3l commented on src/particles/wakefields/WakePush.cpp:
```diff
@@ -0,0 +1,100 @@
+        using namespace amrex::literals;
+
+#if (defined(AMREX_DEBUG) || defined(DEBUG)) && !defined(AMREX_USE_GPU)
+        int const cw_size = convoluted_wakefield.size(); // no padding anymore
```
```suggestion
        int const cw_size = convolved_wakefield.size(); // no padding anymore
```

ax3l commented on src/particles/wakefields/WakePush.cpp:
```diff
@@ -0,0 +1,100 @@
+                amrex::ParticleReal const push_consts = 1.0 / ((ablastr::constant::SI::c) * pz_ref_SI);
+
+                // Gather particles and push momentum
+                const amrex::Real* wakefield_ptr = convoluted_wakefield.data();
```
```suggestion
                const amrex::Real* wakefield_ptr = convolved_wakefield.data();
```

ax3l commented on src/particles/wakefields/WakePush.cpp:
```diff
@@ -0,0 +1,100 @@
+                    amrex::ParticleReal const & AMREX_RESTRICT t = part_t[i];
+                    amrex::ParticleReal & AMREX_RESTRICT pt = part_pt[i];
+
+                    // Update longitudinal momentum with the convoluted wakefield force
```
```suggestion
                    // Update longitudinal momentum with the convolved wakefield force
```

ax3l committed:
Fix Typo: Convolved

Co-authored-by: Chad Mitchell <46825199+cemitch99@users.noreply.github.com>

ax3l commented:
The tolerance appears a little to strict now.

cemitch99 reviewed (approved):
This looks great.  Thanks, Alex!
